### PR TITLE
Use both getrandom() and /dev/urandom by default on Linux.

### PR DESCRIPTION
**Pull request by kroeckx**
getrandom() is now used on Linux by default when using Linux >= 3.17 and glibc >= 2.25

### CODE REVIEW

File: crypto/rand/rand_unix.c [hunk 1]
```diff
@@ -123,13 +119,23 @@ size_t RAND_POOL_acquire_entropy(RAND_POOL *pool)
...
 #  if defined(OPENSSL_RAND_SEED_OS)
-#   if defined(DEVRANDOM)
-#    define OPENSSL_RAND_SEED_DEVRANDOM
-#   else
+#   if !defined(DEVRANDOM)
 #    error "OS seeding requires DEVRANDOM to be configured"
 #   endif
+#   define OPENSSL_RAND_SEED_DEVRANDOM
+#   if defined(__GLIBC__)
+#    if defined(__GLIBC_PREREQ)
```
Comment (richsalz) [inline crypto/rand/rand_unix.c]: The triply-nested if could be collapsed into one.  Side note.

Comment (kroeckx) [reply crypto/rand/rand_unix.c]: Not if you look at the comment in rand_lib.c where I based this on.

Comment (paulidale) [reply crypto/rand/rand_unix.c]: The `defined(__GLIBC__)` and `defined(__GLIBC_PREREQ)` could be combined into a single #if.  The `__GLIBC_PREREQ(2, 25)` cannot and has to stay separate.  rand_lib.c needed to do something special if `*not* defined(__GLIBC__)` which isn't the case here.

File: crypto/rand/rand_unix.c [hunk 2]
```diff
@@ -123,13 +119,23 @@ size_t RAND_POOL_acquire_entropy(RAND_POOL *pool)
...
+#     if __GLIBC_PREREQ(2, 25)
+#      define OPENSSL_RAND_SEED_GETRANDOM
+#     endif
+#    endif
+#   endif
 #  endif
 
+#ifdef OPENSSL_RAND_SEED_GETRANDOM
+# include <sys/random.h>
+#endif
```
Comment (levitte) [inline crypto/rand/rand_unix.c]: Indentation is a bit off...

**Comment by kroeckx** [comment]
I just pushed a new version with those changes.

**Comment by mspncp** [comment]
I think we need a better documentation of the random source selection, otherwise we will get a lot of questions from users in the future, see for example https://github.com/openssl/openssl/issues/5330#issuecomment-364965630.

**Comment by richsalz** [comment]
I agree.  I don't know what Kurt changed, so I'll leave it to him to document.  And think of chroot.  The base stuff I did was all documented.

**Comment by mspncp** [comment]
Since I intend to revise a great part of the random manual pages anyway, I can also add a description of the entropy source selection process.  I will create a WIP pr for the documentation as soon as has reached a certain level of  maturity, so everybody can contribute ideas.

**Comment by richsalz** [comment]
I think in the INSTALL doc, saying that /dev/random might have problems with chroot.
If getrandom is available and works, then the devices aren't consulted, right? So I'll approve.

**Pull request closed without merging**